src/Layer/LabelLayer.js

```diff
@@ -0,0 +1,189 @@
+
+                // FIXME: Style management needs improvement, see #1318.
+                // NOTE: the config param is empty, but will be filled in the
+                // next version.
+                const label = new Label(content, coord.clone(), g.properties.style || f.style || this.style);
+                label.layer = label;
+
+                // coord.x = (coord.x / data.scale.x) - data.translation.x;
+                // coord.y = (coord.y / data.scale.y) - data.translation.y;
+                coord.as(this.crs, coord);
+                coord.toVector3(label.position);
+                label.needsToBeMoved = true;
+
+                if (f.size == 2) {
+                    label.needsAltitude = true;
+                }
+
+                labels.push(label);
+            });
+        });
+
+        return labels;
+    }
+
+    update(context, layer, node, parent) {
+        if (!parent && node.children.length) {
+            // if node has been removed dispose three.js resource
+            ObjectRemovalHelper.removeChildrenAndCleanupRecursively(this, node);
+            return;
+        }
+
+        if (this.frozen || !node.visible) {
+            return;
+        }
+
+        if (node.layerUpdateState[this.id] === undefined) {
+            node.layerUpdateState[this.id] = new LayerUpdateState();
+        }
+
+        if (!node.layerUpdateState[this.id].canTryUpdate()) {
+            return;
+        }
+
+        const extentsDestination = node.getExtentsByProjection(this.source.projection) || [node.extent];
+
+        const extentsSource = [];
+        for (const extentDest of extentsDestination) {
+            const ext = this.source.projection == extentDest.crs ? extentDest : extentDest.as(this.source.projection);
+            if (extentInsideSource(ext, this.source)) {
+                node.layerUpdateState[this.id].noMoreUpdatePossible();
+                return;
+            }
+            extentsSource.push(extentDest);
+        }
+        node.layerUpdateState[this.id].newTry();
+
+        const command = {
+            layer: this,
+            extentsSource,
+            view: context.view,
+            threejsLayer: this.threejsLayer,
+            requester: node,
+        };
+
+        return context.scheduler.execute(command).then((result) => {
+            if (!result) { return; }
+
+            result.forEach((labels) => {
+                if (!node.parent) {
+                    labels.forEach(l => ObjectRemovalHelper.removeChildrenAndCleanupRecursively(this, l));
+                    return;
+                }
+
+                labels.forEach((label) => {
+                    if (label.parent) { return; }
+
+                    label.minlevel = node.extent.zoom;
+
+                    if (label.needsToBeMoved) {
+                        node.worldToLocal(label.position);
+                        label.needsToBeMoved = false;
+                    }
```

Comment:
to remove `needsToBeMoved` 

```suggestion
if (label.needsAltitude) {
    // TODO: add a Style property for the offset
    label.coordinates.z = DEMUtils.getElevationValueAt(this.parent, label.coordinates);
}
label.coord.as(layer.crs, label.position);
node.worldToLocal(label.position);
```

Reply:
It won't work, as `worldToLocal` will be called everytime we add back a `Label` to the scene. Also, it applies a conversion everytime too.

Reply:
I'm not sure because label is added to node, only one time, isn't it?

Reply:
Not if we unzoom and the node is deleted, and we zoom again, the node is added back and we go again here.

Reply:
but if the node is deleted a new label is instanced, no?

Reply:
No, it is only hidden. In my first thoughts, it was costing less to keep a hidden label instead of creating one everytime. But if there is a lot of them, it can be costly indeed. I'll think of a solution, but it seems that adding a `removed` handler will do the trick.

Reply:
In fact, as it is parsed data, it automatically goes into cache (see `DataSourceProvider`). So I'm not sure on what to do.

Reply:
Brainstorm a better solution

Reply:
we're putting aside for the moment